simulation/core/create_index.py

```diff
@@ -0,0 +1,83 @@
+# Copyright 2023 MosaicML Streaming authors
+# SPDX-License-Identifier: Apache-2.0
+
+"""Create a dataset index file from input parameters."""
+
+import os.path
+import sys
+
+sys.path.append(os.path.join(os.path.dirname(__file__), '..'))
+
+import json
+import os
+import random
+import string
+from typing import Optional
+
+
+def get_random_foldername():
+    """Generate random folder name to store the index file in."""
+    return ''.join(
+        random.choice(string.ascii_uppercase + string.ascii_lowercase + string.digits)
+        for _ in range(16))
+
+
+def create_stream_index(shards: int, samples_per_shard: int, avg_raw_shard_size: int,
+                        avg_zip_shard_size: Optional[int]) -> str:
+    """Create dataset index file from input parameters.
+
+    Args:
+        shards (int): Number of shards.
+        samples_per_shard (int): Number of samples per shard.
+        avg_raw_shard_size (int): Average raw shard size.
+        avg_zip_shard_size (int): Average compressed shard size.
+
+    Returns:
+        local path to created index file for stream.
+    """
+    index_data = {'version': 2, 'shards': []}
+
+    shards_list = []
+    for shard_id in range(shards):
+        shard_data = {
+            'column_encodings': [],
+            'column_names': [],
+            'column_sizes': [],
+            'format': 'mds',
+            'raw_data': {
+                'basename': 'shard.' + str(shard_id) + '.mds',
+                'bytes': avg_raw_shard_size,
+                'hashes': {}
+            },
+            'hashes': [],
+            'samples': samples_per_shard,
+            'size_limit': avg_raw_shard_size,
+            'version': 2,
+            'zip_data': None,
+            'compression': None
+        }
+        if avg_zip_shard_size is not None:
+            shard_data['zip_data'] = {
+                'basename': 'shard.' + str(shard_id) + '.mds.zstd',
+                'bytes': avg_zip_shard_size,
+                'hashes': {}
+            }
+            shard_data['compression'] = 'zstd:16'
+        shards_list.append(shard_data)
+
+    index_data['shards'] = shards_list
+
+    # Try making the directory for the stream's index.json file
+    foldername = get_random_foldername() + '_indexcreated'
+    try:
+        os.mkdir(foldername)
+    except FileExistsError:
+        print('Folder already exists, trying again...')
+        foldername = get_random_foldername()
+        os.mkdir(foldername)
```

Comment:
What happens if there is a collision here? Do we want to loop through this logic until we find a unique directory? Maybe try 100/1000 times or so?

Reply:
So the folders get generated and deleted right after they are processed, during `SimulationDataset` instantiation. So there are only a few of these folders that exist at a time. The `get_random_foldername()` function returns a random alphanumeric string of length 16. characters can be lowercase, uppercase, and numeric, so that's 26+26+10 = 62 choices for each character, so overall 62^16 ~= 4.7e28 possible filenames, so the collision probability is super super low. I just added this `FileExistsError` in the very slim off-chance there's a collision, but if you'd want I can increase the number of tries?